### ISSUE TITLE
S3 naming Bug

### ISSUE THREAD

**Issue by KDederichs** [fre5h/VichUploaderSerializationBundle#10]
Hi,

I have a problem with vich uploader, gaufrette and the VichUploaderSerializationBundle and I'm not really sure what is causing it.

I set vich to upload images to an S3 bucket using gaufrette which is working fine.
When I serialize my entity using the JMS Serializer the VichUploaderSerializationBundle will return the link to the S3 bucket.
All that is working fine except that on some seemingly random occasions it will append the S3 url to the database entry for the file name (it even will append it multiple times causeing things like UrlUrlUrlName...)

I have no clue what might be the cause for that because I don't update the entity at all.

Here is part of my configuration:

```
vich_uploader:
    db_driver: orm
    storage: gaufrette
    mappings:
        product_image:
            uri_prefix: 'https://s3-%aws_region%.amazonaws.com/%s3_bucket%/products' # you'll need this set to use the Vich URL generator
            upload_destination: aws_s3_products
            namer: vich_uploader.namer_uniqid
            delete_on_remove: true
            delete_on_update: true

knp_gaufrette:
    stream_wrapper: ~
    adapters:
        products:
            aws_s3:
                service_id: 'aws.s3'
                bucket_name: '%s3_bucket%'
                options:
                    directory: 'products'
                    acl: 'public-read'
    filesystems:
        aws_s3_products:
            adapter:    products
```
Vich part from product entity:

```
/**
     * @ORM\Column(type="string", nullable=true)
     * @Fresh\VichSerializableField("imageFile", includeHost=false)
     * @Expose
     * @Groups({"default"})
     */
    private $imageName;

    /**
     * NOTE: This is not a mapped field of entity metadata, just a simple property.
     *
     * @Assert\File(
     *     maxSize="5M",
     *     mimeTypes={"image/png", "image/jpeg", "image/pjpeg"}
     * )
     * @Vich\UploadableField(mapping="product_image", fileNameProperty="imageName")
     * @Expose
     *
     * @var File
     */
    private $imageFile;
```
I really have no clue what's causing this strange behavior...

I already asked on the vich upload bundle but they said the root of the issue is more likely to be found here.

Thanks in advance

**Comment by KDederichs** [comment]
Just an afterthought, is it maybe possible that the reflection messes things up? 
Like if the doctrine object is flushed and immediately after serialized is there a chance that the reflection happens before the flush action is complete?

**Comment by fre5h** [comment]
Hi. What version of PHP do you use? And do you have this error in production environment only or in dev env also?

**Comment by KDederichs** [comment]
Hi, we use 7.0.13+cli and it is affection our production environment at the moment.

**Comment by fre5h** [comment]
I have not used this bundle with PHP 7.0+ in production. Tests are passed for PHP 7.0 in Travis. But in my projects still PHP 5.6. So I didn't have same problem as you.
So what can it be. Maybe different way of reflection in PHP 7.0. Or maybe some problem with cached annotations. If you use cache for annotations then try to clear it after bug appeared and check if it has gone . Anyway I need some workflow how to reproduce this issue

**Comment by KDederichs** [comment]
I'll try to make one you can replicate:

First of the issue didn't appear for like 2 month so I assume it only appears if you either have a bit of data in your DB or if there is more load on the servers so operations need a tiny bit longer.

The updating of the product, which has the images, happens in controllers that persist and flush around 4 or 5 entities and immediately after serialize an entity which contains a product.

The general flow is:

I persist 4 entities A,B,C,D.

Entity A has a reference to Entity B.
Entity B has a cascade persist annotation which updates an entity (E) that contains my Product.

Immediately after the flush I serialize entity A with a serialization group that exposes entities B and E.

**Comment by fre5h** [comment]
So if you can reproduce this issue, have you debug it locally? Put a breakpoint here https://github.com/fre5h/VichUploaderSerializationBundle/blob/master/EventListener/JmsPreSerializeListener.php#L125 And see URL is duplicated

**Comment by KDederichs** [comment]
So I debugged it at that line and when there is only the filename in the database the uri looks like
https://s3..../name.png
when you run the script again it will be
https://s3..../https://s3....../name.png
and so on.

**Comment by fre5h** [comment]
What I can say exactly, it is not related to S3. I think it is a cache problem

**Comment by KDederichs** [comment]
We deleted our whole symfony cache folder though and the problem still occurred.

**Comment by KDederichs** [comment]
![screenshot from 2017-03-23 13-10-35](https://cloud.githubusercontent.com/assets/24696606/24247410/e6ff06ae-0fcb-11e7-8dbc-b10f81dea01a.png)

Here is a screenshot from the debugger if that helps, the thing I want to persist is the order but somehow it will also persist the imagename of the product

**Comment by fre5h** [comment]
@KDederichs Do you have some fresh information about this issue? Is it still actual and reproducible?

**Comment by KDederichs** [comment]
It should be reproducible like this:

You have some entity that has the VichSerializableField annotation (Entity A) and some entity with a relation to it you are editing at the moment (Entity B).

Now what you are doing is you persist Entity B and directly after that you serialize it (in my case because you want to send it somewhere to announce it).
What will happen since you use reflection to overwrite the original data of Entity A with the FQDN it will pickup on these changes and persist them as well, causing an endless appending of the URL.

What I ended up doing to 'fix' it is I wrote a virtual property containing my URL with placeholders which I then replace with my final data in an Event subscriber. Since it's a virtual property it's not mapped and can't be persisted back into my database.
